ts/component/tab-bar-builder.ts

```diff
@@ -0,0 +1,129 @@
+        onSelect(this.table);
+      }
+    };
+    this.tabContainer = button;
+    this.styleButton();
+  }
+
+  // modifies the tab styling according to whether or not it has been selected
+  public styleButton() {
+    if (this.selected) {
+      this.tabContainer.classList.replace("text-black", "text-green-200");
+      this.tabContainer.classList.add(
+        "ring-2",
+        "ring-green-200",
+        "ring-offset-1"
+      );
+    } else {
+      this.tabContainer.classList.replace("text-green-200", "text-black");
+      this.tabContainer.classList.remove(
+        "ring-2",
+        "ring-green-200",
+        "ring-offset-1"
+      );
+      this.tabContainer.classList.add(
+        "hover:ring-2",
+        "ring-green-200",
+        "ring-offset-1"
+      );
+    }
+  }
+}
+
+export class TabBar {
+  tables: Pair<any, string>[]; // where the key is the table definition, and the value is the table title
+  defaultTable: string;
+  tabBarContainer: HTMLElement;
+  tableContainer: HTMLElement;
+  tabs: Tab[] = [];
+
+  constructor(
+    tables: Pair<any, string>[],
+    defaultTable: string,
+    tabBarContainerId: string,
+    tableContainerId: string
+  ) {
+    // get all containers
+    const tabBarContainer = document.getElementById(tabBarContainerId);
+    if (tabBarContainer === null) {
+      throw Error(`Container ID "${tabBarContainerId}" not found on page`);
+    }
+    this.tabBarContainer = tabBarContainer;
+    const tableContainer = document.getElementById(tableContainerId);
+    if (tableContainer === null) {
+      throw Error(`Container ID "${tableContainerId}" not found on page`);
+    }
+    this.tableContainer = tableContainer;
+    this.defaultTable = defaultTable;
+    this.tables = tables;
+  }
+
+  public build() {
+    const controlsContainer = document.createElement("span");
+    controlsContainer.className =
+      "flex-auto space-x-2 rounded-md px-2 py-2 bg-grey-100 inline-block";
+    // given all the tables and their titles, create the tab bar
+    this.tables.forEach((table, idx) => {
+      const reload = (tb: Pair<any, string>) => {
+        this.reloadTable(tb);
+      };
+      this.tabs.push(
+        new Tab(table, reload, table.value === this.defaultTable ? true : false)
+      );
+      controlsContainer.appendChild(this.tabs[idx].tabContainer);
+    });
+
+    this.tabBarContainer.append(controlsContainer);
+    // build the default table
+    this.tables.forEach((tb) => {
+      if (tb.value === this.defaultTable) {
+        this.reloadTable(tb);
+      }
+    });
+  }
+
+  public reloadTable(table: Pair<any, string>) {
+    this.tabs.forEach((tab) => {
+      if (tab.table.value === table.value) {
+        tab.selected = true;
+        // destroy current table and construct new table
+        this.tableContainer.innerHTML = "";
+        new TableBuilder(
+          table.key,
+          this.tableContainer.id,
+          getSearchParams(),
+          updateUrlParams
+        ).build();
```

Comment:
Move this logic to `detail.ts`. Instead of `tables`, the `TabBar` constructor can take a `Pair<string, () => void>[]`. This way `TabBar` is only responsible for its own behaviour and calling the function defined by its owner. In our current use case, that is rebuilding a table, but it could be used for other things in the future

Reply:
I've modified the tables array to be Pair<Pair<any, string>, () => void>, as when rebuilding the table we'll need the table definition as well. So for example, for the "Cases" table, its pair would look like Pair(Pair(caseDefinition, "Cases"), reload).